### ISSUE TITLE
add pruning

### ISSUE THREAD

**Issue by lwthatcher** [lwthatcher/gfmm#8]
Most of the papers I've referenced state that doing post-pruning can increase generalization.
One of the questions to consider is how often/when to do pruning.
## When to do pruning:
- after each epoch
- after training
- after mini-batch?
- on-line? [hard]

## Progress:
- [ ] optional validation set
- [ ] implement confidence factor function
- [ ] add pruning threshold parameter
- [ ] implement pruning function
- [ ] add parameter to specify how often to prune
